### PR TITLE
Fix wildcard handling of global attributes not present in metadata.

### LINKED ISSUE #14063: This does not play nicely with wildcard reads.  Need to fix ZAP to
put it in the metadata, or fix wildcard expansion to add it.

---

https://github.com/project-chip/connectedhomeip/blob/f0f5de0d79ae1c51d7fddbf70900ebac0acd899a/src/app/util/ember-compatibility-functions.cpp#L307-L312

---

###### This issue was generated by [todo-issue](https://github.com/DerJuulsn/todo-issue) based on a `TODO` comment in f0f5de0d79ae1c51d7fddbf70900ebac0acd899a.

### PR DESCRIPTION
**Pull request by bzbarsky-apple**
Wildcard expansion was not including AttributeList,
ServerGeneratedCommandList, ClientGeneratedCommandList.

Also, AttributeList was not including ServerGeneratedCommandList and
ClientGeneratedCommandList.

Fixes https://github.com/project-chip/connectedhomeip/issues/14063

One caveat: When doing wildcard expansion, this puts the global list attributes at the end, not sorted by id.  I think that's fine for the types of use cases attribute wildcards are used for.

#### Problem
See above.

#### Change overview
See above.

#### Testing
YAML tests are testing that the AttributeList value now includes the command lists.

Manually tested with `chip-tool any read-by-id 6 0xffffffff 17 1` that we now get:
```
[1644467745435] [34678:14913310] CHIP: [TOO] Endpoint: 1 Cluster: 0x0000_0006 Attribute 0x0000_FFF8
[1644467745435] [34678:14913310] CHIP: [TOO]   ServerGeneratedCommandList: 0 entries
[1644467745435] [34678:14913310] CHIP: [TOO] Endpoint: 1 Cluster: 0x0000_0006 Attribute 0x0000_FFF9
[1644467745435] [34678:14913310] CHIP: [TOO]   ClientGeneratedCommandList: 6 entries
[1644467745435] [34678:14913310] CHIP: [TOO]     [1]: 0
[1644467745435] [34678:14913310] CHIP: [TOO]     [2]: 1
[1644467745435] [34678:14913310] CHIP: [TOO]     [3]: 2
[1644467745435] [34678:14913310] CHIP: [TOO]     [4]: 64
[1644467745435] [34678:14913310] CHIP: [TOO]     [5]: 65
[1644467745435] [34678:14913310] CHIP: [TOO]     [6]: 66
[1644467745435] [34678:14913310] CHIP: [TOO] Endpoint: 1 Cluster: 0x0000_0006 Attribute 0x0000_FFFB
[1644467745435] [34678:14913310] CHIP: [TOO]   AttributeList: 10 entries
[1644467745435] [34678:14913310] CHIP: [TOO]     [1]: 0
[1644467745435] [34678:14913310] CHIP: [TOO]     [2]: 16384
[1644467745435] [34678:14913310] CHIP: [TOO]     [3]: 16385
[1644467745435] [34678:14913310] CHIP: [TOO]     [4]: 16386
[1644467745435] [34678:14913310] CHIP: [TOO]     [5]: 16387
[1644467745435] [34678:14913310] CHIP: [TOO]     [6]: 65528
[1644467745435] [34678:14913310] CHIP: [TOO]     [7]: 65529
[1644467745435] [34678:14913310] CHIP: [TOO]     [8]: 65531
[1644467745435] [34678:14913310] CHIP: [TOO]     [9]: 65532
[1644467745435] [34678:14913310] CHIP: [TOO]     [10]: 65533
```
which we did not use to do.

### CODE REVIEW

**Comment by erjiaqing** [comment]
Hint: You need to update `src/app/tests/TestAttributePathExpandIterator.cpp` to pass unit tests.

**Comment by bzbarsky-apple** [comment]
> Hint: You need to update src/app/tests/TestAttributePathExpandIterator.cpp to pass unit tests.

Yep, and some other unit tests are making assumptions about this stuff too.  Will get them sorted.

File: src/app/AttributePathExpandIterator.cpp [hunk 1]
```diff
@@ -101,15 +106,17 @@ void AttributePathExpandIterator::PrepareAttributeIndexRange(const ClusterInfo &
...
+        mGlobalAttributeIndex = 0;
     }
     else
     {
         mAttributeIndex = emberAfGetServerAttributeIndexByAttributeId(aEndpointId, aClusterId, aClusterInfo.mAttributeId);
         // If the given attribute id does not exist on the given endpoint, it will return uint16(0xFFFF), then endAttributeIndex
         // will be 0, means we should iterate a null attribute set (skip it).
-        mEndAttributeIndex = static_cast<uint16_t>(mAttributeIndex + 1);
+        mEndAttributeIndex    = static_cast<uint16_t>(mAttributeIndex + 1);
+        mGlobalAttributeIndex = ArraySize(GlobalAttributesNotInMetadata);
```
Comment (erjiaqing) [inline src/app/AttributePathExpandIterator.cpp:119]: What if the request path is a global attribute?

IIRC reading Wildcard cluster id with global attribute id is legal.

Comment (bzbarsky-apple) [reply src/app/AttributePathExpandIterator.cpp:119]: Good catch!  Let me sort that out.

**Review by erjiaqing — APPROVED** [review]

**Review by jmartinez-silabs — APPROVED** [review]

File: src/app/util/ember-compatibility-functions.cpp [hunk 2]
```diff
@@ -346,29 +347,39 @@ class GlobalAttributeReader : public MandatoryGlobalAttributeReader
...
-            bool addedOurId             = false;
+            const size_t count     = mCluster->attributeCount;
+            bool addedExtraGlobals = false;
             for (size_t i = 0; i < count; ++i)
             {
-                AttributeId id = mCluster->attributes[i].attributeId;
-                if (!addedOurId && id > ourId)
+                AttributeId id              = mCluster->attributes[i].attributeId;
+                constexpr auto lastGlobalId = GlobalAttributesNotInMetadata[ArraySize(GlobalAttributesNotInMetadata) - 1];
+                // We are relying on GlobalAttributesNotInMetadata not having
```
Comment (jmartinez-silabs) [inline src/app/util/ember-compatibility-functions.cpp:360]: Should we add a to-do or git issue to track this?

Comment (bzbarsky-apple) [reply src/app/util/ember-compatibility-functions.cpp:360]: https://github.com/project-chip/connectedhomeip/issues/15011 tracks it.

**Pull request merged**